spring-web/src/main/java/org/springframework/http/server/reactive/JettyCoreServerHttpRequest.java

```diff
@@ -0,0 +1,234 @@
+
+import org.springframework.core.io.buffer.DataBuffer;
+import org.springframework.core.io.buffer.DataBufferFactory;
+import org.springframework.core.io.buffer.PooledDataBuffer;
+import org.springframework.http.HttpCookie;
+import org.springframework.http.HttpHeaders;
+import org.springframework.http.HttpMethod;
+import org.springframework.http.server.RequestPath;
+import org.springframework.http.support.JettyHeadersAdapter;
+import org.springframework.lang.Nullable;
+import org.springframework.util.CollectionUtils;
+import org.springframework.util.LinkedMultiValueMap;
+import org.springframework.util.MultiValueMap;
+import org.springframework.util.StringUtils;
+
+import static org.springframework.http.server.reactive.AbstractServerHttpRequest.QUERY_PATTERN;
+
+/**
+ * Adapt an Eclipse Jetty {@link Request} to a {@link org.springframework.http.server.ServerHttpRequest}.
+ *
+ * @author Greg Wilkins
+ * @since 6.2
+ */
+// TODO: extend AbstractServerHttpRequest for websocket.
+class JettyCoreServerHttpRequest implements ServerHttpRequest {
+	private static final MultiValueMap<String, String> EMPTY_QUERY = CollectionUtils.unmodifiableMultiValueMap(new LinkedMultiValueMap<>());
+
+	private static final MultiValueMap<String, HttpCookie> EMPTY_COOKIES = CollectionUtils.unmodifiableMultiValueMap(new LinkedMultiValueMap<>());
+
+	private final DataBufferFactory dataBufferFactory;
+
+	private final Request request;
+
+	private final HttpHeaders headers;
+
+	private final RequestPath path;
+
+	@Nullable
+	private URI uri;
+
+	@Nullable
+	MultiValueMap<String, String> queryParameters;
+
+	@Nullable
+	private MultiValueMap<String, HttpCookie> cookies;
+
+	public JettyCoreServerHttpRequest(DataBufferFactory dataBufferFactory, Request request) {
+		this.dataBufferFactory = dataBufferFactory;
+		this.request = request;
+		this.headers = new HttpHeaders(new JettyHeadersAdapter(request.getHeaders()));
+		this.path = RequestPath.parse(request.getHttpURI().getPath(), request.getContext().getContextPath());
+	}
+
+	@Override
+	public HttpHeaders getHeaders() {
+		return this.headers;
+	}
+
+	@Override
+	public HttpMethod getMethod() {
+		return HttpMethod.valueOf(this.request.getMethod());
+	}
+
+	@Override
+	public URI getURI() {
+		if (this.uri == null) {
+			this.uri = this.request.getHttpURI().toURI();
+		}
+		return this.uri;
+	}
+
+	@Override
+	public Flux<DataBuffer> getBody() {
+		// We access the request body as a Flow.Publisher, which is wrapped as an org.reactivestreams.Publisher and
+		// then wrapped as a Flux.   The chunks are converted to RetainedDataBuffers with wrapping and can be
+		// retained within a call to onNext.
+
+		// TODO find a better way to release after each onNext call to a subscriber
+		DataBufferReleaser releaser = new DataBufferReleaser();
+		return Flux.from(FlowAdapters.toPublisher(Content.Source.asPublisher(this.request)))
+				.map(this::wrap)
+				.doOnNext(releaser::onNext)
+				.doOnComplete(releaser::onComplete);
```

Comment:
@lachlan-roberts @sbordet  previous comment still applies to this version.  This is a little better, but holds onto a databuffer until the next one arrives, which by definition prevents the previous buffer being reused.  We really want to release after onNext has been called.

Reply:
@lachlan-roberts fixed it!